python/pyarrow/tests/test_pandas.py

```diff
@@ -3158,6 +3158,79 @@ def test_array_protocol():
         assert result.equals(expected2)
 
 
+# ----------------------------------------------------------------------
+# Pandas ExtensionArray support
+
+
+def _Int64Dtype__from_arrow__(self, array):
+    # for test only deal with single chunk for now
+    # TODO: do we require handling of chunked arrays in the protocol?
+    arr = array.chunk(0)
+    buflist = arr.buffers()
+    data = np.frombuffer(buflist[-1], dtype=arr.type.to_pandas_dtype())[
+        arr.offset:arr.offset + len(arr)]
+    bitmask = buflist[0]
+    if bitmask is not None:
+        mask = pa.BooleanArray.from_buffers(
+            pa.bool_(), len(arr), [None, bitmask])
+        mask = np.asarray(mask)
+    else:
+        mask = np.ones(len(arr), dtype=bool)
+    int_arr = pd.arrays.IntegerArray(data.copy(), ~mask, copy=False)
+    return int_arr
+
+
+def test_convert_to_extension_array():
+    if LooseVersion(pd.__version__) < "0.26.0.dev":
+        pytest.skip("Conversion from IntegerArray to arrow not yet supported")
+
+    import pandas.core.internals as _int
+
+    # table converted from dataframe with extension types (so pandas_metadata
+    # has this information)
+    df = pd.DataFrame(
+        {'a': [1, 2, 3], 'b': pd.array([2, 3, 4], dtype='Int64'),
+         'c': [4, 5, 6]})
+    table = pa.table(df)
+
+    # Int64Dtype has no __arrow_array__ -> use normal conversion
+    result = table.to_pandas()
+    assert len(result._data.blocks) == 1
+    assert isinstance(result._data.blocks[0], _int.IntBlock)
+
+    # raise error is explicitly asking for unsupported conversion
+    with pytest.raises(ValueError):
+        table.to_pandas(extension_columns=['b'])
+
+    try:
+        # patch pandas Int64Dtype to have the protocol method
+        pd.Int64Dtype.__from_arrow__ = _Int64Dtype__from_arrow__
```

Comment:
Aside: is there a monkey-patch context manager somewhere?

Reply:
pytest has a `monkeypatch` fixture that we can use. This would look like

```
def test_convert_to_extension_array(monkeypatch):
    ...
    monkeypatch.setattr(pd.Int64Dtype, "__from_arrow__", _Int64Dtype__from_arrow__, raising=False)
    ....
```

without the need for the try/except to ensure clean-up